web/packages/teleterm/src/ui/Search/SearchContext.tsx

```diff
@@ -109,28 +103,58 @@ export const SearchContextProvider: FC = props => {
     setIsOpen(true);
   }
 
+  const [isUserInteractionPaused, setIsUserInteractionPaused] = useState(false);
   /**
-   * lockOpen forces the search bar to stay open for the duration of the action. It also restores
-   * focus on the search input after the action is done.
+   * pauseUserInteraction temporarily causes addWindowEventListener not to add listeners for the
+   * duration of the action. It also restores focus on the search input after the action is done.
    *
-   * This is useful in situations where want the search bar to not close when the user interacts
-   * with modals shown from the search bar.
+   * This is useful in situations where want the search bar to show some other element the user can
+   * interact with, for example a modal. When the user interacts with the modal, we don't want the
+   * search bar listeners to intercept those interactions, which could for example cause the search
+   * bar to close or make the user unable to press Enter in the modal as the search bar would
+   * swallow that event.
    */
-  async function lockOpen(action: Promise<any>): Promise<void> {
-    setIsLockedOpen(true);
-
-    try {
-      await action;
-    } finally {
-      // By the time the action passes, the user might have caused the focus to be lost on the
-      // search input, so let's bring it back.
-      //
-      // focus needs to be executed before the state update, otherwise the search bar will close for
-      // some reason.
-      inputRef.current?.focus();
-      setIsLockedOpen(false);
-    }
-  }
+  const pauseUserInteraction = useCallback(
+    async (action: () => Promise<any>): Promise<void> => {
+      setIsUserInteractionPaused(true);
+
+      try {
+        await action();
+      } finally {
+        // By the time the action passes, the user might have caused the focus to be lost on the
+        // search input, so let's bring it back.
+        //
+        // focus needs to be executed before the state update, otherwise the search bar will close
+        // for some reason.
+        inputRef.current?.focus();
+        setIsUserInteractionPaused(false);
+      }
+    },
+    []
+  );
+
+  /**
+   * addWindowEventListener is meant to be used in useEffect calls which register event listeners
+   * related to the search bar. It automatically removes the listener when the user interaction gets
+   * paused.
+   *
+   * pauseUserInteraction is supposed to be called in situations where the search bar is obstructed
+   * by another element (such as a modal) and we don't want interactions with that other element to
+   * have any effect on the search bar.
+   */
+  const addWindowEventListener = useCallback(
+    (...args: Parameters<typeof window.addEventListener>) => {
+      if (isUserInteractionPaused) {
+        return;
+      }
+
+      window.addEventListener(...args);
+      return () => {
```

Comment:
I think we had a similar situation in the past, IMO it is more readable when the returned function is named:
```ts
return {
  cleanup: () => window.removeListenter....
}
```
EDIT: I realized that the retuned function is used as a cleanup function function in `useEffect` automatically. Soo..., idk we can leave it as is, or call the function explicitly. Personally, I like the second option a little more (by being more explicit), but I don't have a strong opinion.

Reply:
In either case the biggest problem that persists is that there's no way to prevent someone from just forgetting about returning the cleanup from `useEffect`. I guess a custom ESLint rule would be nice but no one has time to work on this atm.

I'll see how it looks with an explicit `cleanup` return, as you said it'll at least make it more explicit in existing callsites that `addWindowEventListener` returns a cleanup function.